### PR TITLE
[Inductor] Async Pipelined Autotune

### PR DESCRIPTION
**Pull request by PaulZhang12**
Async-Pipelined Autotuning is a path for Inductor max-autotune-gemm to do autotuning in a subprocess during lowering time, overlapping autotuning with lowering/scheduling time in parallel to reduce compilation overhead. Below, we highlight the base case versus this new path:

**Baseline**
1. Go through lowering path for each gemm, calling precompile synchronously for each choice and returning
2. During scheduling time, `get_min_choice` for each MultiTemplateBuffer does the actual autotuning synchronously. 

However, this approach results in the GPU being idle up until the autotuning time for each choice is actually needed. Async-Pipelined Autotuning addresses this issue by being able to kick off autotuning for each choice async as soon as possible, to attempt to keep the GPU fed throughout the compilation process for max-autotune-gemm.

**Async-Pipelined Autotuning**
1. Compile on main process and thread, precompilation thread pool on main process, 1 autotuning subprocess
2. Precompilation is now truly async, for each gemm that is lowered, do kick off "starting precompilation" completely async in a separate thread pool and return immediately
3. Kick off autotuning in subprocess for extern choices before precompilation of triton choices, want to keep GPU subprocess fed
4. After precompilation is done, separate thread invokes the autotune subprocess to tune that precompiled choice

Below is a diagram of the general architecture with Diode, the ML GEMM model that can be plugged into Inductor:

image

**E2E numbers**
Based on a portion of a representative internal model:
Base: ~14s compilation time for graph
Topk=1 Diode + disable epilogue fusion: ~30s compilation time
Topk=1 Diode + disable epilogue fusion + Async lower: ~27s compilation time

**TODO**

* Support Prologue/Epilogue Fusion
* Support Subgraph Choice Callers

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/_inductor/autotune_process.py [hunk 1]
```diff
@@ -1050,3 +1056,270 @@ def benchmark_in_sub_process(
...
+
+
+def _warmup_autotune_subprocess() -> bool:
+    """
+    Warmup function run in the autotune subprocess.
+    """
+    import torch
+
+    # Initialize dummy tensor for CUDA context
+    if torch.cuda.is_available():
```
Comment (masnesral) [inline torch/_inductor/autotune_process.py:1177]: mostly out of curiosity: Given our discussion in chat, I wouldn't think this is necessary. We believe that importing torch is enough to init CUDA?

File: torch/_inductor/autotune_process.py [hunk 2]
```diff
@@ -1050,3 +1056,270 @@ def benchmark_in_sub_process(
...
+
+        Returns:
+            Dict mapping ChoiceCaller to benchmark timing
+        """
+
+        timings = {}
+        for choice in choices:
+            choice_hash = AsyncAutotuner.get_choice_hash(choice, inputs_key)
+            timings[choice] = AsyncAutotuner.choice_hash_to_future[choice_hash].result(
+                timeout=timeout
```
Comment (masnesral) [inline torch/_inductor/autotune_process.py]: Do you want to del AsyncAutotuner.choice_hash_to_future[choice_hash] here?

File: torch/_inductor/config.py [hunk 3]
```diff
@@ -466,6 +466,11 @@ def prologue_fusion_enabled() -> bool:
     os.environ.get("TORCHINDUCTOR_DISTRIBUTED_MAX_AUTOTUNE_GEMM") == "1"
 )
 
+# Pipeline autotuning for max-autotune-gemm. Overlap lowering and benchmarking on GPU
+pipeline_max_autotune_gemm = (
```
Comment (masnesral) [inline torch/_inductor/config.py:465]: To make internal rollout easier, you probably want to create a JK. Here's an example of how we do that elsewhere in this file: https://fburl.com/code/y8ez9f7e

File: torch/_inductor/select_algorithm.py [hunk 4]
```diff
@@ -2830,40 +2831,93 @@ def __call__(
...
 
         if return_multi_template and (config.max_autotune or config.max_autotune_gemm):
-
-            def get_timings(hint_override: Optional[int] = None):
-                filtered_choices = [
-                    c
-                    for c in choices
-                    if not hasattr(c, "hint_override")
-                    or c.hint_override == hint_override
+            if config.pipeline_max_autotune_gemm:
```
Comment (masnesral) [inline torch/_inductor/select_algorithm.py:2835]: nit: just for purposes of readability, is it possible to split this out into two different helpers: one for the pipelined impl and one for the non-pipelined impl? The deep nesting is getting a bit much.

File: torch/_inductor/autotune_process.py [hunk 5]
```diff
@@ -1050,3 +1056,270 @@ def benchmark_in_sub_process(
...
+        Uses ProcessPoolExecutor with 'spawn' context for CUDA safety.
+        ProcessPoolExecutor is lazily initialized - workers are not spawned
+        until the first submit() call, making this property non-blocking.
+        """
+        if self._pool is None:
+            with self._lock:
+                if self._pool is None:
+                    # Use 'spawn' context to avoid CUDA fork issues
+                    # Workers are spawned lazily on first submit(), not here
+                    ctx = mp.get_context("spawn")
```
Comment (masnesral) [inline torch/_inductor/autotune_process.py]: How good is the processpoolexecutor about cleaning up subprocs if the parent process doesn't exit cleanly (like a crash in cpp). I don't really know the history, but our other process pools feel the need to do this stuff: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/compile_worker/utils.py#L25

File: torch/_inductor/autotune_process.py [hunk 6]
```diff
@@ -1050,3 +1056,270 @@ def benchmark_in_sub_process(
...
+    _lock: threading.Lock = threading.Lock()
+
+    def __init__(self, num_workers: int = 1):
+        self._num_workers = num_workers
+        self._pool: ProcessPoolExecutor | None = None
+        self._warmup_future: Future[Any] | None = None
+        self._warmup_start_time: float | None = None
+
+    @classmethod
+    def get_instance(cls, num_workers: int = 1):
```
Comment (masnesral) [inline torch/_inductor/autotune_process.py]: nit: The API is kinda weird because if you call get_instance() multiple times with a different number of workers, you get the old value right? Which is just to say that maybe it doesn't need to be configurable?

File: torch/_inductor/autotune_process.py [hunk 7]
```diff
@@ -1050,3 +1056,270 @@ def benchmark_in_sub_process(
...
+    """
+
+    _instance: Optional[PrecompileThreadPool] = None
+    _lock = threading.Lock()
+
+    def __init__(self, max_workers: int = 4):
+        self._executor = ThreadPoolExecutor(max_workers=max_workers)
+
+    @classmethod
+    def get_instance(cls, num_workers: int = 4) -> PrecompileThreadPool:
```
Comment (masnesral) [inline torch/_inductor/autotune_process.py]: nit: The API is weird here too for the same reason. Maybe the get_num_workers() logic from select_algorithm.py should just move here and the num_workers shouldn't pretend to be configurable? Just a suggestion. I don't feel strongly; in practice it would be strange if the num_workers was changing within the process.

File: torch/_inductor/select_algorithm.py [hunk 8]
```diff
@@ -2830,40 +2831,93 @@ def __call__(
...
+                # Extern kernels do not have to precompile, so can feed them before triton
+                AsyncAutotuner.start(extern_kernels, inputs_key)
+                triton_kernels = [
+                    c for c in choices if not AlgorithmSelectorCache._is_extern(c)
                 ]
-                timings = self.do_autotuning(
+                precompile_instance = PrecompileThreadPool.get_instance(
+                    num_workers=get_num_workers()
+                )
+                precompile_future = precompile_instance.submit(
```
Comment (masnesral) [inline torch/_inductor/select_algorithm.py:2852]: Also left a msg in chat. I can't quite figure out what's going on here. The existing code is already so confusing :/  Is this thread pool actually being used to precompile anything? I _think_ the precompile_fn at this point is just a function to wait on all precompile futures submitted to the async_compile pool. This "precompile_fn" is just waiting for the futures to finish.

File: torch/_inductor/config.py (hunk 3 above)
Comment (PaulZhang12) [reply torch/_inductor/config.py:465]: @masnesral yeah, this is enabled off by default for now and I intend to enable it on a case by case basis, no wide rollout at all currently

File: torch/_inductor/autotune_process.py (hunk 5 above)
Comment (PaulZhang12) [reply torch/_inductor/autotune_process.py]: Hmm I am not sure here, I know that other ProcessPoolExecutors are used in Inductor

File: torch/_inductor/autotune_process.py (hunk 1 above)
Comment (PaulZhang12) [reply torch/_inductor/autotune_process.py:1177]: Will test!

File: torch/_inductor/autotune_process.py (hunk 2 above)
Comment (PaulZhang12) [reply torch/_inductor/autotune_process.py]: Actually I think this is missing some code here, which is basic caching logic (not reautotuning on choices that have been autotuned before). Will add that back in, but this allows us to nicely store the result for future autotune choices

File: torch/_inductor/select_algorithm.py (hunk 8 above)
Comment (PaulZhang12) [reply torch/_inductor/select_algorithm.py:2852]: Responded with this in chat:
"This PR basically makes it so that the do_autotuning is wrapped in a future completely. That way, when the precompilation is complete for a single triton choice, another thread will send it to the autotune subprocess as soon as possible"

File: torch/_inductor/config.py (hunk 3 above)
Comment (masnesral) [reply torch/_inductor/config.py:465]: I meant if you just create the JK now, it provides you a convenient way to do that rollout.

File: torch/_inductor/autotune_process.py (hunk 5 above)
Comment (masnesral) [reply torch/_inductor/autotune_process.py]: Maybe it's fine; something to keep in mind if you see old processes sticking around.

**Review by masnesral — APPROVED** [review]

File: torch/_inductor/autotune_process.py (hunk 1 above)
Comment (PaulZhang12) [reply torch/_inductor/autotune_process.py:1177]: It seems like you actually need this to initialize cuda, which makes sense. I believe in the past I've seen torch lazy init cuda

File: torch/_inductor/autotune_process.py [hunk 9]
```diff
@@ -1050,3 +1056,270 @@ def benchmark_in_sub_process(
...
+        if cls._instance is None:
+            with cls._lock:
+                if cls._instance is None:
+                    cls._instance = cls(num_workers)
+        return cls._instance
```
Comment (nmacchioni) [inline torch/_inductor/autotune_process.py:1236]: Copy-pasting my comment from the last PR since it is still relevant:

This doesn't seem right, originally each call to precompile would get its own thread pool with 1 thread per 1 choice (assuming max # threads is larger). Now each call to precompile shares the same thread pool, which has # of threads = to the # of choices on the first call to precompile? So not only do subsequent precompile calls have to wait for prior precompile calls to free up threads, but we might never have enough threads to process an entire precompile at once (if the first call has less choices than subsequent calls)

**Review by nmacchioni — COMMENTED** [review]
What benefit do you see in having PrecompileThreadPool rather than a regular ThreadPoolExecutor in the init of AlgorithmSelectorCache?

File: torch/_inductor/autotune_process.py (hunk 9 above)
Comment (PaulZhang12) [reply torch/_inductor/autotune_process.py:1236]: I see ~5% faster compile time from not constructing a threadpool for every choice

File: torch/_inductor/autotune_process.py (hunk 2 above)
Comment (PaulZhang12) [reply torch/_inductor/autotune_process.py]: @masnesral oh I misread the code actually, the caching logic is there. In start, we have `choice_hash in AsyncAutotuner.choice_hash_to_future`, and if it's true, we don't submit a new task to the process group

File: torch/_inductor/config.py (hunk 3 above)
Comment (PaulZhang12) [reply torch/_inductor/config.py:465]: Yeah I understand, I have no intention of doing any such rollout in the short term future :)

File: torch/_inductor/select_algorithm.py (hunk 4 above)
Comment (PaulZhang12) [reply torch/_inductor/select_algorithm.py:2835]: I'm not sure this will actually be better... we depend on quite a few local variables and we would just be passing a lot of them around anyways

File: torch/_inductor/autotune_process.py (hunk 9 above)
Comment (nmacchioni) [reply torch/_inductor/autotune_process.py:1236]: threadpool per choice or thread per choice?

Comment (PaulZhang12) [reply torch/_inductor/autotune_process.py:1236]: Threadpool per shape -> single thread for every choice. Basically the default. Reconstructing this thread pool in another thread pool generally seems bad for perf

Comment (nmacchioni) [reply torch/_inductor/autotune_process.py:1236]: Sure, I agree that creating a new threadpool per call is not ideal. I'm just saying that the logic for setting the max_workers is not consistent with the prior implementation.

Previously, one threadpool per shape meant one thread for every choice (in the common case).

Now, one threadpool for all shapes has some # of max_workers dependent on who initialized the threadpool. So now we'll only have ~20 (avg # of configs) threads floating around for all of the jobs in total. So let's say instead of top-1 pruning you used top-10 or something. The threadpool would be initialized with 11 (10 triton + 1 aten) threads at the first call. Any subsequent calls would submit jobs to this threadpool, but would only be queued since the 11 threads are already in use. So you'd end up getting a backup.

File: torch/_inductor/autotune_process.py [hunk 10]
```diff
@@ -1050,3 +1056,261 @@ def benchmark_in_sub_process(
     Do benchmarking in a subprocess and return the perf number (latency).
     """
     return get_tuning_process_pool().benchmark(choices)
+
+
+class AutotuneProcessPool:
+    """
+    Singleton pool manager for running autotuning (precompilation + benchmarking)
```
Comment (eellison) [inline torch/_inductor/autotune_process.py:1063]: Is this intended that we're running precompilation ? ideally this would be handled separately. cc @masnesral

File: torch/_inductor/autotune_process.py [hunk 11]
```diff
@@ -1050,3 +1056,263 @@ def benchmark_in_sub_process(
...
+        """
+        Get or create the process pool.
+
+        Uses ProcessPoolExecutor with 'spawn' context for CUDA safety.
+        ProcessPoolExecutor is lazily initialized - workers are not spawned
+        until the first submit() call, making this property non-blocking.
+        """
+        # Use 'spawn' context to avoid CUDA fork issues
+        # Workers are spawned lazily on first submit(), not here
+        ctx = mp.get_context("spawn")
```
Comment (eellison) [inline torch/_inductor/autotune_process.py:1102]: deferring to @masnesral on spawn issues

File: torch/_inductor/scheduler.py [hunk 12]
```diff
@@ -2847,6 +2847,9 @@ def _init(self, nodes: list[ir.Operation]) -> None:
 
         self.merge_loops()
         self.finalize_multi_template_buffers()
+        if config.max_autotune_gemm or config.max_autotune:
+            torch._inductor.select_algorithm.PrecompileThreadPool.shutdown_instance()
```
Comment (eellison) [inline torch/_inductor/scheduler.py:2853]: what is the overhead of a single instantiation? would it make sense to say startup on fwd, close on corresponding bwd?

File: torch/_inductor/select_algorithm.py [hunk 13]
```diff
@@ -3300,7 +3357,11 @@ def on_complete(future):
                     elapsed_seconds,
                 )
 
-        executor = ThreadPoolExecutor(max_workers=num_workers)
```
Comment (eellison) [inline torch/_inductor/select_algorithm.py:3304]: note: the executor is only conditionally used, so im not sure this change does what you want..

File: torch/_inductor/select_algorithm.py [hunk 14]
```diff
@@ -3394,8 +3455,6 @@ def wait_on_futures():
             if exceptions:
                 _log_autotune_exceptions(exceptions)
 
-            executor.shutdown(wait=True)
```
Comment (eellison) [inline torch/_inductor/select_algorithm.py:3398]: we should still shutdown right? int he case of the existing executor

**Review by eellison — APPROVED** [review]
looks good - mostly deferring to @masnesral

File: torch/_inductor/select_algorithm.py (hunk 13 above)
Comment (PaulZhang12) [reply torch/_inductor/select_algorithm.py:3304]: executor is conditionally used, but in the base case we initialize it anyways?

File: torch/_inductor/select_algorithm.py (hunk 14 above)
Comment (PaulZhang12) [reply torch/_inductor/select_algorithm.py:3398]: Yes that's right, I forgot to push another change

File: torch/_inductor/autotune_process.py (hunk 11 above)
Comment (masnesral) [reply torch/_inductor/autotune_process.py:1102]: Yeah, I think the use of spawn wouldn't be acceptable in the long term if this feature is to be used in OSS. I think we could piggy-back on top of the implementation for Triton compilation, but @PaulZhang12 ran into issues around CUDA initialization. I feel like there's a way to fix that, but I think the thinking here was: let's get this landed so Paul can test it internally and iterate. Enabling for OSS down the road would need a different solution w.r.t. spawn. @eellison chime in if you feel like this should work cleanly in OSS too before landing.

**Comment by PaulZhang12** [comment]
Got reverted in OSS while landed internally due to lint error. Redelegating to https://github.com/pytorch/pytorch/pull/170407
